Add support flash driver for STM32F7x SoC

jinyongiot opened the pull request:
This patch add support flash driver for STM32F7x SoC. So the boards of STM32F7x SoC can use the flash_stm32 interface. Test on stm32f769i_disco

aurel32 commented:
I have also been working on flash support for STM32F7 during the week-end, and was about to submit it. My branch is available there: https://github.com/aurel32/zephyr/tree/stm32f7-flash . That way I know exactly what to look for, so I will review your changes.

aurel32 commented on drivers/flash/flash_stm32.c:
```diff
@@ -104,6 +107,14 @@ static void flash_stm32_flush_caches(struct device *dev)
+		regs->acr.val &= ~FLASH_ACR_ARTRST;
+		regs->acr.val |= FLASH_ACR_ARTEN;
+	}
+#endif
```
This is only valid for access through the ITCM bus. Usually the code is executed through the AXIS bus, and goes through the cache. The data cache should be invalidated through `SCB_InvalidateDCache_by_Addr`. I don't think we should flush the instruction cache after an erase. See https://github.com/aurel32/zephyr/commit/0188becc6d3e91933b6536509a49fe9953fe2acd

aurel32 commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,201 @@
+	regs->cr |= FLASH_CR_PG;
+
+	/* flush the register write */
+	tmp = regs->cr;
```
The reference manual recommends to use a DSB instruction to flush the register. The Cortex-M7 is using a more advanced pipeline, so the trick of doing a dummy read might not be enough.

aurel32 commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,201 @@
+	tmp = regs->cr;
+
+	*((u8_t *) offset + CONFIG_FLASH_BASE_ADDRESS) = val;
+
```
You should insert a DSB instruction as recommended in the reference manual. I actually have seen issues on the STM32F723 where some data when not written to the flash.

aurel32 commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,201 @@
+
+	/* flush the register write */
+	tmp = regs->cr;
+
```
Same comment about adding a DSB instruction here.

aurel32 commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,201 @@
+ * sectors (16 KB) in the second bank occurring after the large ones
+ * (128 KB) in the first. We could consider supporting this as two
+ * devices to make the layout cleaner, but this will do for now.
+ */
```
That is not true for the STM32F769 which can be configured in two different way, with either 12 sectors (4*32K, 128K, 7*256K) or 24 sectors (twice 4*16K, 64K, 7*128K).

aurel32 commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,201 @@
+	{.pages_count = 1, .pages_size = KB(128)},
+	{.pages_count = 3, .pages_size = KB(256)},
+};
+#elif FLASH_SECTOR_TOTAL == 12
```
Note that this case never appears in the STM32 HAL.

aurel32 commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,201 @@
+	{.pages_count = 1, .pages_size = KB(128)},
+	{.pages_count = 7, .pages_size = KB(256)},
+};
+#elif FLASH_SECTOR_TOTAL == 16
```
Same here, this case never appears in the STM32 HAL.

aurel32 commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,201 @@
+	{.pages_count = 4, .pages_size = KB(16)},
+	{.pages_count = 1, .pages_size = KB(64)},
+	{.pages_count = 7, .pages_size = KB(128)},
+};
```
As said above, there are two cases depending of the `nDBANK` bit of the `FLASH_OPTCR` register. You could have a look how I have done that there: https://github.com/aurel32/zephyr/commit/7b6e90dd019dd68cee463c40230f23aa0fdbd81b#diff-322955a6f2ebc21dba22c2c642a8b3faR129

aurel32 commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,201 @@
+};
+#else
+#error "Unknown flash layout"
+#endif /* FLASH_SECTOR_TOTAL == 8 */
```
Could you please add the other STM32F7 SoCs? You could get that part there: https://github.com/aurel32/zephyr/commit/7b6e90dd019dd68cee463c40230f23aa0fdbd81b#diff-322955a6f2ebc21dba22c2c642a8b3faR129

jinyongiot replied:
I am sorry, I don't test my code.
Thank you for pointing out the error of my code.
Now, I rewrite my code refer to your code, but test no pass on my board stm32f769i_disco too.
So, I want to write it refer to the STM32 HAL.

aurel32 reviewed (changes requested):
Could you please tell use what you tested with this flash driver? At least for me I was not able to do a sector erase without getting a timeout error. It can be quite long on the STM32F7 (up to 4 seconds for the biggest sectors). I have a patch for that there: 
https://github.com/aurel32/zephyr/commit/10d3541262fcc02e9922ba800db1237716afb873

superna9999 reviewed (changes requested):
I'm ok with @aurel32 comments, please fix these issues and explicit how and on with SoC you did the tests

jinyongiot commented:
Modify the code and test it time after time. It can run on the stm32f769i_disco board, now. The tested code path is samples / drivers / flash_shell.

aurel32 commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,206 @@
+	regs->cr &= STM32F7X_SECTOR_MASK;
+	regs->cr |= FLASH_CR_SER | (sector << FLASH_CR_SNB_Pos);
+	regs->cr |= FLASH_CR_STRT;
+
```
Similarly to what is done for writing a byte, you should set the `PSIZE` field to `BYTE`. It's possible to get a bigger erase parallelism on the STM32F7, but it is not possible over the whole supported voltage range. Given we do not know the voltage, it should be forced to byte.

I believe this can be improved later by adding a new entry in the device tree, but that should be done later in an another patch.

aurel32 commented on drivers/flash/Kconfig.stm32:
```diff
@@ -10,16 +10,18 @@ if FLASH && SOC_FAMILY_STM32
 	select FLASH_HAS_PAGE_LAYOUT if SOC_SERIES_STM32L4X
 	help
-	 Enable STM32F0x, STM32F3x, STM32F4x OR STM32L4x series flash driver.
+	 Enable STM32F0x, STM32F3x, STM32F4x STM32F7x OR STM32L4x series flash driver.
```
Minor things, there is a missing comma after STM32F4x

aurel32 reviewed (changes requested):
Thanks for this new version, there are only minor things left. I have tried it on my STM32F723 discovery board using an NVS filesystem and confirm it all works fine.

jinyongiot commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,206 @@
+	if (sector > 11) {
+		sector += 4;
+	}
+
```
In dual bank mode, the SNB msb is selects the bank2, others select the sector in bank2.  The sector 12 is the sector 0 in bank2, so set the SNB to 0x10H (16D), this RM0410 page 110 say.

aurel32 reviewed (commented):
Thanks for the update, most of the changes are fine, we are almost there. There are only a few minor things left around `STM32F7X_SECTOR_MASK`.

jinyongiot commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,213 @@
+
+#if FLASH_SECTOR_TOTAL == 24
+	/* Dual bank mode, SNB MSB selects the bank2, others select sector. */
+	if (sector > 11) {
```
By 1Mbytes devices, the sector8, sector9, sector10, sector11 are invalid. the first sector on bank2 is sector12, same as 2Mbytes devices. (table 6, RM0410)

aurel32 replied:
You are correct, I had the impression that we have to account for the missing sectors 7-11 for 1MB devices, but this will have to be done at the flash page layout level. Anyway those devices are not supported and the flash page layout API does not support defining holes in the map.

jinyongiot commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,213 @@
+
+	/* clear PSIZE bits and SNB bits */
+	regs->cr &= STM32F7X_SECTOR_MASK;
+	regs->cr |= FLASH_PSIZE_BYTE;
```
you are right! 
And I like the first more.

aurel32 reviewed (approved):
Thanks @jinyongiot for your work, from my point of view this is ready to be merged. @superna9999, do you still have some comments about this PR?

aurel32 commented:
recheck

jinyongiot commented:
My suggestion: For stm32f7x 1Mbytes devices, we can use a sector number map to the devices sector number. Example, use sector8 map to the devices sector12 (first sector on the bank2).

superna9999 commented on drivers/flash/flash_stm32.c:
```diff
@@ -257,7 +272,7 @@ static int stm32_flash_init(struct device *dev)
 	k_sem_init(&p->sem, 1, 1);
 
-	return flash_stm32_write_protection(dev, false);
+	return flash_stm32_write_protection(dev, true);
```
What is the reason to change this here ? Is this related with F7 support ?

jinyongiot replied:
I just think that set flash protection false in init api is not necessary.

superna9999 replied:
Please push in a separate commit and a separate PR, this is not the right place to change that

erwango commented on drivers/flash/flash_stm32.c:
```diff
@@ -14,7 +14,19 @@
 #include "flash_stm32.h"
 
-#define STM32_FLASH_TIMEOUT	((u32_t) 0x000B0000)
+/* STM32F0: maximum erase time of 40ms for a 2K sector */
```
For better clarity, can you put the comments inside the ifdef's ?

erwango commented on drivers/flash/flash_stm32.c:
```diff
@@ -63,49 +75,50 @@ static int flash_stm32_check_status(struct device *dev)
 	}
-}
+#elif defined(CONFIG_SOC_SERIES_STM32F7X)
+	SCB_InvalidateDCache_by_Addr((void *)CONFIG_FLASH_BASE_ADDRESS
```
ARG_UNUSED(dev); ?

erwango commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
```
Remove this one and put your own copyright here.

erwango commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,249 @@
+#define STM32F7X_SECTOR_MASK		((u32_t) 0xFFFFFF07)
+
+bool flash_stm32_valid_range(struct device *dev, off_t offset, u32_t len,
+			     bool write)
```
Align 'bool' on 'struct' on at upper line

erwango commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,249 @@
+	if (sector > 11) {
+		sector += 4;
+	}
+#elif CONFIG_FLASH_SIZE == 1024
```
You have same value in #if and #elif, is that intentional?

erwango commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,249 @@
+}
+
+int flash_stm32_block_erase_loop(struct device *dev, unsigned int offset,
+				 unsigned int len)
```
align 'unsigned' on 'struct' in upper line

erwango commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,249 @@
+}
+
+int flash_stm32_write_range(struct device *dev, unsigned int offset,
+			    const void *data, unsigned int len)
```
Align 'const' on 'struct' in upper line

erwango commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,249 @@
+	return rc;
+}
+
+/* Different SoC flash layouts are specified in across various
```
remove 'across'

erwango commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,249 @@
+
+/* Different SoC flash layouts are specified in across various
+ * reference manuals, but the flash layout for a given number of
+ * sectors. In case of 8 sectors we need to differentiate between
```
'but the flash layout for a given number of sectors.' Do we miss a word in this part of the sentence?

erwango commented on drivers/flash/flash_stm32f7x.c:
```diff
@@ -0,0 +1,249 @@
+
+#endif /* !defined(FLASH_OPTCR_nDBANK) */
+void flash_stm32_page_layout(struct device *dev,
+			     const struct flash_pages_layout **layout,
```
Align 'const' on 'struct'

erwango reviewed (changes requested):
Some comments but good work otherwise! Thanks

jinyongiot committed:
driver: flash: add flash driver support for STM32F7x SoC

Change little flash_stm32 interface code for support STM32F7x SoC,
and add STM32F7x SoC flash_stm32 interface driver.

Signed-off-by: Yong Jin <jinyong.iot@foxmail.com>

jinyongiot committed:
dts: arm: stm32f7: add flash controller support for STM32F7x SoC

Add support for flash controller support for STM32F7x SoC in dts.

Signed-off-by: Yong Jin <jinyong.iot@foxmail.com>

erwango reviewed (changes requested):
Please take into account the comments about function header alignment.
Otherwise looks good

erwango commented:
@jinyongiot , any update?

erwango commented:
Closing as replaced by #12083
